File: go-port/pkg/workspace/layout.go
package workspace

import "github.com/notepad-plus-plus/ultra-project/pkg/core"

// Deep comment: Workspace manages the UI layout state, specifically fulfilling the Vertical Tabs requirement.
// Why: Legacy Notepad++ ties tabs tightly to the Win32 TabCtrl. By extracting this into a Go model,
//      we can instruct the UI (Qt/GTK) to render tabs however we want—specifically, as a vertical
//      list with adjustable column widths rather than rotated 90-degree text.
// Side effects: The UI layer must poll or listen to changes in this model to know which buffer is focused.

type TabPlacement string

const (
	PlacementTop      TabPlacement = "TOP"
	PlacementVertical TabPlacement = "VERTICAL"
)

type Tab struct {
	BufferID core.BufferID
	Title    string
	IsActive bool
}

type Layout struct {
	Placement       TabPlacement
	VerticalWidthPx int
	Tabs            []*Tab
}

func NewLayout() *Layout {
	return &Layout{
		Placement:       PlacementVertical, // Fulfilling user requirement as default
		VerticalWidthPx: 250,
		Tabs:            make([]*Tab, 0),
	}
}

func (l *Layout) AddTab(bufID core.BufferID, title string) {
	// Deactivate existing
	for _, t := range l.Tabs {
		t.IsActive = false
	}

	// Add new active tab
	l.Tabs = append(l.Tabs, &Tab{
		BufferID: bufID,
		Title:    title,
		IsActive: true,
	})
}

func (l *Layout) SetVerticalWidth(width int) {
	if width > 50 && width < 1000 {
		l.VerticalWidthPx = width
	}
}

File: rust-port/src/workspace/layout.rs
use parking_lot::RwLock;
use crate::core::buffer::BufferId;
use std::sync::Arc;

#[derive(Clone, PartialEq, Debug)]
pub enum TabPlacement {
    Top,
    Vertical,
}

#[derive(Clone, Debug)]
pub struct Tab {
    pub buffer_id: BufferId,
    pub title: String,
    pub is_active: bool,
}

pub struct Layout {
    pub placement: Arc<RwLock<TabPlacement>>,
    pub vertical_width_px: Arc<RwLock<u32>>,
    pub tabs: Arc<RwLock<Vec<Tab>>>,
}

impl Layout {
    pub fn new() -> Self {
        Self {
            placement: Arc::new(RwLock::new(TabPlacement::Vertical)), // Fulfilling user requirement as default
            vertical_width_px: Arc::new(RwLock::new(300)),
            tabs: Arc::new(RwLock::new(Vec::new())),
        }
    }

    pub fn add_tab(&self, buf_id: BufferId, title: String) {
        let mut tabs = self.tabs.write();

        for tab in tabs.iter_mut() {
            tab.is_active = false;
        }

        tabs.push(Tab {
            buffer_id: buf_id,
            title,
            is_active: true,
        });
    }

    pub fn set_vertical_width(&self, width: u32) {
        if width > 50 && width < 1000 {
            *self.vertical_width_px.write() = width;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_layout_default_vertical() {
        let layout = Layout::new();
        assert_eq!(*layout.placement.read(), TabPlacement::Vertical);
    }

    #[test]
    fn test_add_tab() {
        let layout = Layout::new();
        layout.add_tab("1".to_string(), "test.txt".to_string());

        let tabs = layout.tabs.read();
        assert_eq!(tabs.len(), 1);
        assert!(tabs[0].is_active);
    }
}
